src/ucs/vfs/fuse/vfs_fuse.c

```diff
@@ -254,6 +254,16 @@ static ucs_status_t ucs_vfs_fuse_wait_for_path(const char *path)
 
     pthread_mutex_lock(&ucs_vfs_fuse_context.mutex);
 
+    /* Check 'stop' flag before entering the loop. If the main thread sets
+     * 'stop' flag before this thread created 'inotify_fd' fd, the execution
+     * of the thread has to be stopped, otherwise - the thread hangs waiting
+     * for the data on 'inotify_fd' fd.
+     */
+    if (ucs_vfs_fuse_context.stop) {
+        status = UCS_ERR_CANCELED;
+        goto out;
+    }
```

Comment:
after moving checking of stop flag here, i think there could be a race, leading to a hang: the main thread will set stop=1 after the VFS thread already passed this point. Then, ucs_vfs_fuse_context.inotify_fd is still -1, so the main thread will not call inotify_rm_watch. Then, this VFS thread will set a watch and never wake up from it.
That's why the previos code checked stop flag only **after** creating a watch.

Reply:
This assumption is wrong.
Bot fuse and main threads use mutex to synchronise access, which means that main thread cannot update the state in between, while fuse thread is holding the lock. They cannot execute mutually exclusive blocks simultaneously, that's the whole point of having mutex
```
fuse thread:
  lock
    check stop
   <<< NO interruption is possible here
    check inotify_fd = inotify_init
  unlock

main thread:
  lock
    stop = 1
    check inotify_fd
  unlock
```

Reply:
ok